[GH-108] Generate Kotlin classes from JSON Schema

kezhenxu94 opened the pull request:
resolve #108 

Generate Kotlin classes from JSON Schema

wuseal commented on src/main/kotlin/wu/seal/jsontokotlin/JsonSchemaDataClassGenerator.kt:
```diff
@@ -0,0 +1,73 @@
+    val requiredFields = jsonObjectDef.required ?: emptyArray()
+    val properties = if (ConfigManager.isOrderByAlphabetical) jsonObjectDef.properties.toSortedMap() else jsonObjectDef.properties
+    val s = TypeSpec.classBuilder(clazz).apply {
+      if (!ConfigManager.isCommentOff && (jsonObjectDef.description?.isNotBlank() == true)) {
```
Here may no need to judge the config , the `isCommentOff`　function was implement by interceptor, so threre is no need to deal with option function here.

wuseal replied:
Just add the comment, no need to check config option

kezhenxu94 replied:
The comment is presented as Kotlin Doc(multiple lines) in this case, and the interceptor did not deal with the Kotlin Doc by now

wuseal replied:
Got it! now need to check if  code parser could work ok with this new type comment
https://github.com/wuseal/JsonToKotlinClass/blob/master/src/main/kotlin/wu/seal/jsontokotlin/utils/classblockparse/ClassCodeParser.kt

kezhenxu94 replied:
@wuseal  I'll try to put this enhancement into the next version since time limits

wuseal commented on src/main/kotlin/wu/seal/jsontokotlin/JsonSchemaDataClassGenerator.kt:
```diff
@@ -0,0 +1,73 @@
+        ?: throw IllegalArgumentException("className cannot be null when jsonObjectDef.title is null")
+
+    val requiredFields = jsonObjectDef.required ?: emptyArray()
+    val properties = if (ConfigManager.isOrderByAlphabetical) jsonObjectDef.properties.toSortedMap() else jsonObjectDef.properties
```
The same as isComment off, `OrderByAlphabetial`　also implemented by interceptors , no need to do it here

wuseal replied:
Just as it original sequence is OK

wuseal commented on src/main/kotlin/wu/seal/jsontokotlin/JsonSchemaDataClassGenerator.kt:
```diff
@@ -0,0 +1,73 @@
+          } else ClassName("", property.capitalize())
+          addProperty(
+              PropertySpec.builder(property, typeName).apply {
+                if (!ConfigManager.isCommentOff && (propertyDefinition.description?.isNotBlank() == true)) {
```
And The same herer about `isCommentOff` option

wuseal commented on src/main/kotlin/wu/seal/jsontokotlin/JsonSchemaDataClassGenerator.kt:
```diff
@@ -0,0 +1,73 @@
+
+          addProperty(
+              PropertySpec.builder(property, typeName).apply {
+                if (!ConfigManager.isCommentOff && (propertyDefinition.description?.isNotBlank() == true)) {
```
And here

kezhenxu94 commented:
Hi. @Rody66,  please refer to [this link](https://help.github.com/en/articles/checking-out-pull-requests-locally) to pull this pr to your local machine, since you are not a maintainer of this repo, please open another pr with your changes, thanks

Rody66 commented:
@kezhenxu94 Ok, let me work on this. Should I commit changes to kezhenxu94:feature/json-schema or somewhere else?

kezhenxu94 commented:
> @kezhenxu94 Ok, let me work on this. Should I commit changes to kezhenxu94:feature/json-schema or somewhere else?

@Rody66 you don't have permission to push to my branch, so please pull my changes locally, add your changes and re-create a pull request here, please note that **the next version is due on this weekend**, so try your best :) before that

kezhenxu94 commented:
I think this PR is ready for merging, I've added basic tests and some advanced usages are in plan to be added in the future.

Rody66 commented:
> 
> 
> I think this PR is ready for merging, I've added basic tests and some advanced usages are in plan to be added in the future.

Hello @kezhenxu94 ! I implemented many of JsonSchema cases, include NJsonSchema `x-*` fields. JsonSchema structure appeared to be really complicated and flexible. My implementation support `$ref` links (to local properties and definitions), combinations (`oneOf`, `allOf`), enums (include `x-enumNames`), multi-type (real nullable types like `"type": ["null", "object"]`), arrays and jagged arrays. So it's able to generate Kotlin data from very complicated schemas. I started work on this two days ago, so my changes can conflict with your last commits. I suggest you to look how we can merge it.

kezhenxu94 commented:
> > I think this PR is ready for merging, I've added basic tests and some advanced usages are in plan to be added in the future.
> 
> Hello @kezhenxu94 ! I implemented many of JsonSchema cases, include NJsonSchema `x-*` fields. JsonSchema structure appeared to be really complicated and flexible. My implementation support `$ref` links (to local properties and definitions), combinations (`oneOf`, `allOf`), enums (include `x-enumNames`), multi-type (real nullable types like `"type": ["null", "object"]`), arrays and jagged arrays. So it's able to generate Kotlin data from very complicated schemas. I started work on this two days ago, so my changes can conflict with your last commits. I suggest you to look how we can merge it.

Hi @Rody66 , could you please open another PR that includes your changes? And I'll try to resolve the conflicts

Rody66 commented:
#148
